src/mongo_odm/odm_collection.hpp

```diff
@@ -0,0 +1,787 @@
+* @param opt An optional BSON document view
+* @return   An optional object of type T
+*/
+template <class T>
+mongocxx::stdx::optional<T> to_optional_obj(
+    const mongocxx::stdx::optional<bsoncxx::document::value>& opt) {
+    if (opt) {
+        T obj = to_obj<T>(opt.value().view());
+        return {obj};
+    } else {
+        return {};
+    }
+}
+
+template <class T>
+class deserializing_cursor;
+
+template <class T>
+class serializing_iterator;
+
+template <class T>
+class odm_collection {
+   public:
+    odm_collection(mongocxx::collection& c) noexcept;
+
+    odm_collection(odm_collection&&) noexcept;
+    odm_collection& operator=(odm_collection&&) noexcept;
+    ~odm_collection();
+
+    /**
+     * Returns a copy of the underlying collection.
+     * @return The mongocxx::collection that this object wraps
+     */
+    mongocxx::collection collection();
+
+    ///
+    /// Runs an aggregation framework pipeline against this collection, and returns the results
+    /// as
+    /// de-serialized object.
+    ///
+    /// @param pipeline
+    ///   The pipeline of aggregation operations to perform.
+    /// @param options
+    ///   Optional arguments, see mongocxx::mongocxx::options::aggregate.
+    ///
+    /// @return A deserializing_cursor with the results.
+    /// @throws
+    ///   If the operation failed, the returned cursor will throw an exception::query
+    ///   when it is iterated.
+    ///
+    /// @see http://docs.mongodb.org/manual/reference/command/aggregate/
+    ///
+    deserializing_cursor<T> aggregate(
+        const mongocxx::pipeline& pipeline,
+        const mongocxx::options::aggregate& options = mongocxx::options::aggregate());
+
+    ///
+    /// Counts the number of documents matching the provided filter.
+    ///
+    /// @param filter
+    ///   The filter, as a serializable object, that documents must match in order to be
+    ///   counted.
+    /// @param options
+    ///   Optional arguments, see mongocxx::options::count.
+    ///
+    /// @return The count of the documents that matched the filter.
+    /// @throws exception::query if the count operation fails.
+    ///
+    /// @see http://docs.mongodb.org/manual/reference/command/count/
+    ///
+    std::int64_t count(T filter,
+                       const mongocxx::options::count& options = mongocxx::options::count());
+
+    ///
+    /// Deletes all matching documents from the collection.
+    ///
+    /// @param filter
+    ///   A serializable object representing the data to be deleted.
+    /// @param options
+    ///   Optional arguments, see mongocxx::mongocxx::options::delete_options.
+    ///
+    /// @return The optional result of performing the deletion, a result::delete_result.
+    /// @throws exception::write if the delete fails.
+    ///
+    /// @see http://docs.mongodb.org/manual/reference/command/delete/
+    ///
+    mongocxx::stdx::optional<mongocxx::result::delete_result> delete_many(
+        const T& filter,
```

Comment:
In our team discussion earlier, did we agree to remove these overloads in favor of BSON only filtering?